Empty-body response from opencode server surfaces as opaque 'Unexpected end of JSON input'

## Summary

**Updated root cause - this is NOT about reasoning models.**

Any condition that causes the opencode server to return an empty 200 response body surfaces to consumers as `SyntaxError: Unexpected end of JSON input` with no actionable context. The most common trigger is an invalid or unavailable model ID, but the same failure mode applies to any empty-body response class.

The original hypothesis (gpt-5 emits `reasoning` parts that `extractTextFromParts` drops) is incorrect. The error happens before `doGenerate` ever reaches the parts-extraction code - it crashes inside `@opencode-ai/sdk` while parsing the HTTP response body.

## Real stack (captured with local instrumentation)

```
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:5880:19)
    at successSteps (node:internal/deps/undici/undici:5861:27)
    at fullyReadBody (node:internal/deps/undici/undici:4753:9)
    at consumeBody (node:internal/deps/undici/undici:5870:7)
    at request (@opencode-ai/sdk/dist/gen/client/client.gen.js:83:28)
    at OpencodeLanguageModel.doGenerate (ai-sdk-provider-opencode-sdk/dist/index.js:1130:22)
```

`@opencode-ai/sdk` line 83 blindly calls `response.json()` regardless of body presence. When the body is empty (which opencode server returns for conditions like unknown model IDs), undici's `JSON.parse('')` throws.

## Reproduce

Any invalid model ID triggers it:

```
# Direct CLI - tells you exactly what's wrong:
$ opencode run --model openai/gpt-5.2 "test"
Error: Model not found: openai/gpt-5.2

$ opencode run --model github-copilot/gpt-5 "test"
Error: Model not found: github-copilot/gpt-5. Did you mean: gpt-5.2, gpt-5.4, gpt-5.1?

# Same models via this SDK - cryptic JSON error:
$ task-master models --set-main openai/gpt-5.2 --opencode
$ task-master parse-prd prd.md --num-tasks=5
[ERROR] OpenCode object generation failed: Unexpected end of JSON input
```

## Triggers confirmed

- `openai/gpt-5.2` with no direct OpenAI auth configured in opencode (provider prefix invalid for the user's setup)
- `github-copilot/gpt-5` (the real Copilot model is `gpt-5.2`, not `gpt-5`)
- Any other unknown provider/model combination

`github-copilot/gpt-5.2` and `github-copilot/gpt-4.1` work fine - the models resolve, body comes back populated, no crash.

## Why the original hypothesis was wrong

The empty-body failure happens at the HTTP layer inside `@opencode-ai/sdk`, before `doGenerate` even gets response data. The `extractTextFromParts` code was a red herring because the request never reaches it - it throws during the initial `client.session.prompt()` call.

## Impact

Developer UX only. When a user misconfigures the model ID (easy to do given the `provider/model` format and the fact that Copilot's gpt-5 family is named `gpt-5.2`, not `gpt-5`), they get an opaque JSON parse error instead of "model not found". This surfaces downstream as `AI_APICallError: Unexpected end of JSON input` with no model context, which is what initially misled me in this issue.

## Proposed fix (in `ai-sdk-provider-opencode-sdk`)

Wrap the `client.session.prompt()` call in `doGenerate` (and the equivalent in `doStream`) with a narrow catch:

```ts
try {
  const result = await client.session.prompt({ path: { id: sessionId }, body: requestBody });
  // ... existing handling
} catch (error) {
  if (error instanceof SyntaxError && /Unexpected end of JSON input/.test(error.message)) {
    throw new APICallError({
      message: `OpenCode returned an empty response body for model "${this.modelId}". This usually means the model is unavailable or the ID is incorrect. Run \`opencode models\` to see valid IDs. (Original: ${error.message})`,
      url: 'opencode://session',
      requestBodyValues: requestBody,
      cause: error,
    });
  }
  throw error;
}
```

This preserves all successful paths and only transforms the specific empty-body symptom into an actionable diagnostic.

## Upstream fix (in `@opencode-ai/sdk`, separate concern)

The deeper fix belongs in the generated client at `client.gen.js:83`: before calling `response.json()`, check `Content-Length === 0` or wrap the call and return `null`/throw a typed error with status code. That surfaces the real HTTP-level problem to all consumers, not just ones via this provider.

Happy to PR the narrow fix here. The upstream fix I'd leave as a separate issue against `@opencode-ai/sdk` / opencode's main repo.

## Environment

- SDK: `ai-sdk-provider-opencode-sdk@0.0.3` (ai-sdk-v5 tag)
- AI SDK: `ai@^5.0.51`
- Node: 20
- OpenCode: 1.4.3
- Shell verified: `opencode run --model github-copilot/gpt-5.2` succeeds; same model via this SDK also works (earlier test in claude-task-master#1685)

Follow-up with a second distinct failure mode observed during the same investigation.

With the **valid** Copilot model (`github-copilot/gpt-5.2`) rather than an invalid one, the error changes:

```
[ERROR] OpenCode object generation failed: terminated
  "data":{"errorType":"TypeError","sessionId":"ses_2731ecf0affeV7eaHUU1EzaqpQ"}
```

That's undici's connection-terminated error - the HTTP request was cut off mid-response. Likely cause: gpt-5.2 is a reasoning model and takes significantly longer than gpt-4.1, so the underlying fetch hits a timeout (node default ~300s, or the `serverTimeout` setting in the provider) before the body finishes streaming back.

Same developer-UX pain as the empty-body case: the error reaches the consumer as `AI_APICallError` with `errorType: TypeError` and message `terminated`, which doesn't tell anyone the actual problem is "your reasoning model is slower than the HTTP timeout".

Suggested handling alongside the `SyntaxError` catch proposed in the issue body:

```ts
if (error instanceof TypeError && /terminated/.test(error.message)) {
  throw new APICallError({
    message: `OpenCode request was terminated mid-response for model "${this.modelId}". This usually happens when a reasoning model takes longer than the HTTP timeout. Bump \`serverTimeout\` in the provider settings. (Original: ${error.message})`,
    url: 'opencode://session',
    requestBodyValues: requestBody,
    cause: error,
  });
}
```

Both catches can live in the same narrow wrapper around `client.session.prompt()`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - **Empty OpenCode response errors** - Empty-body JSON parse failures from the OpenCode server (`SyntaxError: Unexpected end of JSON input`, most commonly caused by an invalid or unavailable `provider/model` ID) are now wrapped in an actionable `APICallError` that names the requested model ID and likely cause. The same wrapping is applied to the streaming prompt failure path, and `modelId` is now included in API call error metadata. (Fixes [#21](https://github.com/ben-vargas/ai-sdk-provider-opencode-sdk/issues/21), PR [#24](https://github.com/ben-vargas/ai-sdk-provider-opencode-sdk/pull/24) by [@slegarraga](https://github.com/slegarraga))
+- **image-input example** - Updated `examples/image-input.ts` to use `openai/gpt-5.5` instead of the no-longer-available `openai/gpt-5.3-codex`, which caused the example to fail silently (empty response, zero usage).
 
 ## [3.0.3] - 2026-06-09
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [3.0.4] - 2026-06-11
+
+### Fixed
+
+- **Empty OpenCode response errors** - Empty-body JSON parse failures from the OpenCode server (`SyntaxError: Unexpected end of JSON input`, most commonly caused by an invalid or unavailable `provider/model` ID) are now wrapped in an actionable `APICallError` that names the requested model ID and likely cause. The same wrapping is applied to the streaming prompt failure path, and `modelId` is now included in API call error metadata. (Fixes [#21](https://github.com/ben-vargas/ai-sdk-provider-opencode-sdk/issues/21), PR [#24](https://github.com/ben-vargas/ai-sdk-provider-opencode-sdk/pull/24) by [@slegarraga](https://github.com/slegarraga))
+
 ## [3.0.3] - 2026-06-09
 
 ### Fixed

--- a/examples/image-input.ts
+++ b/examples/image-input.ts
@@ -50,7 +50,7 @@ async function main() {
     const mediaType = SUPPORTED_EXTENSIONS[ext]!;
 
     const result = streamText({
-      model: opencode("openai/gpt-5.3-codex"),
+      model: opencode("openai/gpt-5.5"),
       messages: [
         {
           role: "user",

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "ai-sdk-provider-opencode-sdk",
-  "version": "3.0.3",
+  "version": "3.0.4",
   "description": "AI SDK v6 provider for OpenCode via @opencode-ai/sdk",
   "keywords": [
     "ai-sdk",

--- a/src/errors.test.ts
+++ b/src/errors.test.ts
@@ -341,6 +341,27 @@ describe("errors", () => {
       expect(result).toBeInstanceOf(APICallError);
     });
 
+    it("should wrap empty JSON responses with model guidance", () => {
+      const error = new SyntaxError("Unexpected end of JSON input");
+      const result = wrapError(error, {
+        sessionId: "session-123",
+        modelId: "github-copilot/gpt-5",
+      });
+
+      expect(result).toBeInstanceOf(APICallError);
+      expect(result.message).toContain(
+        "OpenCode returned an empty JSON response",
+      );
+      expect(result.message).toContain('for model "github-copilot/gpt-5"');
+      expect(result.message).toContain("provider/model");
+      expect((result as APICallError).isRetryable).toBe(false);
+      expect((result as APICallError).data).toMatchObject({
+        errorType: "EmptyJSONResponse",
+        sessionId: "session-123",
+        modelId: "github-copilot/gpt-5",
+      });
+    });
+
     it("should pass metadata through", () => {
       const error = { message: "Error" };
       const result = wrapError(error, { sessionId: "session-123" });

--- a/src/errors.ts
+++ b/src/errors.ts
@@ -6,6 +6,7 @@ import { APICallError, LoadAPIKeyError } from "@ai-sdk/provider";
 export interface OpencodeErrorMetadata {
   sessionId?: string;
   messageId?: string;
+  modelId?: string;
   statusCode?: number;
   isRetryable: boolean;
   errorType?: string;
@@ -175,6 +176,35 @@ export function createAPICallError(
       errorType: metadata?.errorType ?? extractErrorType(error),
       sessionId: metadata?.sessionId,
       messageId: metadata?.messageId,
+      modelId: metadata?.modelId,
+    },
+  });
+}
+
+/**
+ * Create an actionable error for empty JSON responses from OpenCode.
+ */
+function createEmptyResponseError(
+  error: unknown,
+  metadata?: Partial<OpencodeErrorMetadata>,
+): APICallError {
+  const modelHint = metadata?.modelId ? ` for model "${metadata.modelId}"` : "";
+  const originalMessage = extractErrorMessage(error);
+
+  return new APICallError({
+    message:
+      `OpenCode returned an empty JSON response${modelHint}. ` +
+      "This usually means the OpenCode server rejected the request before returning response data, often because the configured provider/model is unavailable or invalid. " +
+      `Check the model ID and OpenCode provider configuration. Original error: ${originalMessage}`,
+    url: "opencode://session",
+    requestBodyValues: {},
+    statusCode: metadata?.statusCode ?? extractStatusCode(error),
+    isRetryable: false,
+    data: {
+      errorType: "EmptyJSONResponse",
+      sessionId: metadata?.sessionId,
+      messageId: metadata?.messageId,
+      modelId: metadata?.modelId,
     },
   });
 }
@@ -357,10 +387,34 @@ export function wrapError(
     return createTimeoutError(timeoutMs, "request");
   }
 
+  if (isEmptyJsonResponseError(error)) {
+    return createEmptyResponseError(error, metadata);
+  }
+
   return createAPICallError(error, metadata);
 }
 
 const DEFAULT_REQUEST_TIMEOUT_MS = 5000;
+
+function isEmptyJsonResponseError(error: unknown): boolean {
+  if (!error || typeof error !== "object") {
+    return false;
+  }
+
+  const err = error as Record<string, unknown>;
+  const name =
+    error instanceof SyntaxError
+      ? "SyntaxError"
+      : typeof err.name === "string"
+        ? err.name
+        : undefined;
+
+  return (
+    name === "SyntaxError" &&
+    typeof err.message === "string" &&
+    err.message === "Unexpected end of JSON input"
+  );
+}
 
 function extractTimeoutMs(error: unknown): number | undefined {
   if (!error || typeof error !== "object") {

--- a/src/opencode-language-model.test.ts
+++ b/src/opencode-language-model.test.ts
@@ -337,6 +337,20 @@ describe("opencode-language-model", () => {
       expect(result.request?.body).toBeDefined();
     });
 
+    it("should wrap empty OpenCode JSON responses with model context", async () => {
+      mockClient.session.prompt.mockRejectedValueOnce(
+        new SyntaxError("Unexpected end of JSON input"),
+      );
+
+      await expect(
+        model.doGenerate({
+          prompt: basicPrompt,
+        }),
+      ).rejects.toThrow(
+        'OpenCode returned an empty JSON response for model "anthropic/claude-3-5-sonnet-20241022"',
+      );
+    });
+
     it("should handle JSON mode", async () => {
       await model.doGenerate({
         prompt: basicPrompt,
@@ -1048,7 +1062,10 @@ describe("opencode-language-model", () => {
     });
 
     it("should emit StructuredOutput tool input as text content", async () => {
-      const structuredInput = { output: "# Hello World", outputType: "markdown" };
+      const structuredInput = {
+        output: "# Hello World",
+        outputType: "markdown",
+      };
       mockClient.session.prompt.mockResolvedValueOnce({
         data: {
           info: {
@@ -1093,10 +1110,14 @@ describe("opencode-language-model", () => {
 
       const textParts = result.content.filter((c) => c.type === "text");
       expect(textParts).toHaveLength(1);
-      expect(JSON.parse((textParts[0] as { text: string }).text)).toEqual(structuredInput);
+      expect(JSON.parse((textParts[0] as { text: string }).text)).toEqual(
+        structuredInput,
+      );
 
       const toolCalls = result.content.filter((c) => c.type === "tool-call");
-      const toolResults = result.content.filter((c) => c.type === "tool-result");
+      const toolResults = result.content.filter(
+        (c) => c.type === "tool-result",
+      );
       expect(toolCalls).toHaveLength(0);
       expect(toolResults).toHaveLength(0);
     });
@@ -1149,7 +1170,9 @@ describe("opencode-language-model", () => {
 
       const textParts = result.content.filter((c) => c.type === "text");
       expect(textParts).toHaveLength(1);
-      expect(JSON.parse((textParts[0] as { text: string }).text)).toEqual({ result: "done" });
+      expect(JSON.parse((textParts[0] as { text: string }).text)).toEqual({
+        result: "done",
+      });
     });
 
     it("should not emit tool-call/tool-result for error-status StructuredOutput", async () => {
@@ -1184,7 +1207,9 @@ describe("opencode-language-model", () => {
       });
 
       const toolCalls = result.content.filter((c) => c.type === "tool-call");
-      const toolResults = result.content.filter((c) => c.type === "tool-result");
+      const toolResults = result.content.filter(
+        (c) => c.type === "tool-result",
+      );
       expect(toolCalls).toHaveLength(0);
       expect(toolResults).toHaveLength(0);
     });

--- a/src/opencode-language-model.ts
+++ b/src/opencode-language-model.ts
@@ -274,11 +274,11 @@ export class OpencodeLanguageModel implements LanguageModelV3 {
       };
 
       this.logger.debug?.(
-        `[doGenerate] raw parts from OpenCode: ${JSON.stringify(responseData.parts?.map(p => ({ type: p.type, ...(p.type === 'text' ? { text: (p as any).text?.slice(0, 200) } : {}), ...(p.type === 'tool' ? { tool: (p as any).tool, callID: (p as any).callID, status: (p as any).state?.status, input: (p as any).state?.input } : {}) })))}`,
+        `[doGenerate] raw parts from OpenCode: ${JSON.stringify(responseData.parts?.map((p) => ({ type: p.type, ...(p.type === "text" ? { text: (p as any).text?.slice(0, 200) } : {}), ...(p.type === "tool" ? { tool: (p as any).tool, callID: (p as any).callID, status: (p as any).state?.status, input: (p as any).state?.input } : {}) })))}`,
       );
       const content = this.extractContentFromParts(responseData.parts ?? []);
       this.logger.debug?.(
-        `[doGenerate] extracted content: ${JSON.stringify(content.map(c => ({ type: c.type, ...(c.type === 'text' ? { text: (c as any).text?.slice(0, 200) } : {}), ...(c.type === 'tool-call' ? { toolName: (c as any).toolName } : {}) })))}`,
+        `[doGenerate] extracted content: ${JSON.stringify(content.map((c) => ({ type: c.type, ...(c.type === "text" ? { text: (c as any).text?.slice(0, 200) } : {}), ...(c.type === "tool-call" ? { toolName: (c as any).toolName } : {}) })))}`,
       );
       const usage = this.extractUsageFromParts(responseData.parts ?? []);
       const finishReason = mapOpencodeFinishReason(responseData.info);
@@ -304,6 +304,7 @@ export class OpencodeLanguageModel implements LanguageModelV3 {
 
       throw wrapError(error, {
         sessionId: this.sessionId,
+        modelId: this.modelId,
       });
     }
   }
@@ -418,8 +419,12 @@ export class OpencodeLanguageModel implements LanguageModelV3 {
           }
 
           client.session.prompt(requestBody).catch((error: unknown) => {
-            logger.error(`Prompt error: ${extractErrorMessage(error)}`);
-            controller.enqueue({ type: "error", error });
+            const wrappedError = wrapError(error, {
+              sessionId,
+              modelId: this.modelId,
+            });
+            logger.error(`Prompt error: ${extractErrorMessage(wrappedError)}`);
+            controller.enqueue({ type: "error", error: wrappedError });
             resolvePromptFailed?.();
           });
 
@@ -521,8 +526,12 @@ export class OpencodeLanguageModel implements LanguageModelV3 {
             controller.enqueue(createStreamStartPart(streamWarnings));
           }
           if (!isAbortError(error)) {
-            logger.error(`Stream error: ${extractErrorMessage(error)}`);
-            controller.enqueue({ type: "error", error: wrapError(error) });
+            const wrappedError = wrapError(error, {
+              sessionId,
+              modelId: this.modelId,
+            });
+            logger.error(`Stream error: ${extractErrorMessage(wrappedError)}`);
+            controller.enqueue({ type: "error", error: wrappedError });
           }
         } finally {
           controller.close();
@@ -794,11 +803,14 @@ export class OpencodeLanguageModel implements LanguageModelV3 {
       // `Output.object()` / `Output.array()` can parse it via `step.text`.
       // Emit the tool input as a text part instead of tool-call/tool-result.
       if (toolPart.tool === STRUCTURED_OUTPUT_TOOL) {
-        const serialized = safeStringifyToolInput(toolPart.state.input, (message) => {
-          this.logger.warn(
-            `Failed to serialize StructuredOutput input for ${toolPart.callID}: ${message}`,
-          );
-        });
+        const serialized = safeStringifyToolInput(
+          toolPart.state.input,
+          (message) => {
+            this.logger.warn(
+              `Failed to serialize StructuredOutput input for ${toolPart.callID}: ${message}`,
+            );
+          },
+        );
         this.logger.debug?.(
           `[StructuredOutput doGenerate] callID=${toolPart.callID} status=${toolPart.state.status} raw input=${JSON.stringify(toolPart.state.input)} serialized=${serialized}`,
         );

